### DISCUSSION TITLE
Using a custom session key for some subdomains

### DISCUSSION THREAD

**Discussion by strika** [rack/rack-session#25] [Q&A]
We're working on a Rails application that uses subdomains. We need to have a different session for different subdomains:

- `app.example.com` should use the "default" session
- `abc.example.com` should use a session that's only for the `abc` domain
- `demo.example.com` should use a session that's only for the `demo` domain
- all other subdomains should use the default session

One potential solution is to change the session key before each request based on the subdomain. The idea is to do that in a middleware.

A similar question was asked before - https://github.com/rack/rack/issues/1155.

The proposed solution was to change the key by accessing the session options:

```ruby
env["rack.session.options"][:key] = "my_cookie_name"
```

However, as the author of the previous question/issue explained:

> The abstract session store implementation ([lib/rack/session/abstract/id.rb](https://github.com/rack/rack/blob/v2.2.8/lib/rack/session/abstract/id.rb#L253)) uses an instance variable, initialized at startup, to determine the name of the cookie it will use to fetch/set a session's id.

Latter, the instance variable is used to fetch the [session ID](https://github.com/rack/rack/blob/v2.2.8/lib/rack/session/abstract/id.rb#L321).

So, changing the session key like this:

```ruby
env["rack.session.options"][:key] = "my_cookie_name"
```

Doesn't have any effect on the instance variable.

At the moment, we're using Rack 2.2.8.

Is there something obvious that we're missing? Thank you.

**Comment by god-of-javascript** [comment]
This is needed, anyone has any idea how we can achieve this? maybe @womblep ?

**Reply by womblep** [reply]
Not I. I havent needed to change this so I am not sure how that part works.
